core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -934,43 +927,32 @@ class KafkaApis(val requestChannel: RequestChannel,
                     OffsetFetchResponse.INVALID_OFFSET, OffsetFetchResponse.NO_METADATA, Errors.forException(e)))
             }
           }.toMap
-          new OffsetFetchResponse(Errors.NONE, (responseInfo ++ unauthorizedStatus).asJava, header.apiVersion)
-        }
-        else {
+
+          val unauthorizedPartitionData = unauthorizedPartitions.map(_ -> OffsetFetchResponse.UNKNOWN_PARTITION).toMap
+          new OffsetFetchResponse(Errors.NONE, (authorizedPartitionData ++ unauthorizedPartitionData).asJava, header.apiVersion)
+        } else {
           // versions 1 and above read offsets from Kafka
-          val offsets = coordinator.handleFetchOffsets(offsetFetchRequest.groupId,
-            if (offsetFetchRequest.isAllPartitions)
-              None
-            else
+          if (offsetFetchRequest.isAllPartitions) {
+            val (error, allPartitionData) = coordinator.handleFetchOffsets(offsetFetchRequest.groupId)
+            if (error != Errors.NONE)
+              offsetFetchRequest.getErrorResponse(error)
+            else {
+              // clients are not allowed to see offsets for topics that are not authorized for Describe
+              val authorizedPartitionData = allPartitionData.filterKeys(authorizeTopicDescribe)
```

Comment:
`filterKeys` (like `mapValues`) returns a view over the original collection. That means that calling `get` on the resulting map would call `authorize` which is not what we want.